### PR TITLE
CmdLine does not forwarding signals

### PR DESCRIPTION
**Pull request by EzzhevNikita**
**Task name**: <Name of changed or new pipeline task>
BashV3, CmdLineV2

**Description**: <Describe your changes here>
Was added sending signals to the task child processes. (Need to trap SIGTERM signal in bash script)

**Documentation changes required:** (Y/N) <Please mark if documentation changes are required>
No
**Added unit tests:** (Y/N) <Please mark if unit tests were added or updated according changes>
No
**Attached related issue:** (Y/N) <Please add link to related issue here>
#11468
**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/BashV3/bash.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ async function run() {
             if (old_source_behavior) {
                 contents = `. '${targetFilePath.replace(/'/g, "'\\''")}' ${input_arguments}`.trim();
             } else {
-                contents = `bash '${targetFilePath.replace(/'/g, "'\\''")}' ${input_arguments}`.trim();
+                contents = `${process.platform == 'win32' ? "" : "exec "}bash '${targetFilePath.replace(/'/g, "'\\''")}' ${input_arguments}`.trim();
```
Comment (anatolybolshakov) [inline Tasks/BashV3/bash.ts:82]: Nit: could you please move conditional logic out of template -  for some readability?

Comment (EzzhevNikita) [reply Tasks/BashV3/bash.ts]: It is not necessary changes so I just removed this condition.

**Comment by anatolybolshakov** [comment]
Please check if we need to apply changes for Powershell handler too

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks!

**Comment by EzzhevNikita** [comment]
This problem is actual only for unix systems as Windows doesn't support POSIX signals and we can't send send specific signals from Bash and CmdLine tasks on windows, so the the changes for powershell handler is unnecessary.

> Please check if we need to apply changes for Powershell handler too

File: Tasks/BashV3/bash.ts [hunk 2]
```diff
@@ -128,6 +128,10 @@ async function run() {
             ignoreReturnCode: true
         };
 
+        process.on("SIGINT", () => {
+            bash.killChildProcess();
```
Comment (alexander-smolyakov) [inline Tasks/BashV3/bash.ts:132]: Could you please add some log-message for this case?

File: Tasks/BashV3/bash.ts [hunk 3]
```diff
@@ -153,6 +157,14 @@ async function run() {
 
         let result = tl.TaskResult.Succeeded;
 
+        /**
+         * Exit code null could appeared in situations if executed script don't process cancellation signal,
+         * as we already have message after operation cancellation, we can avoid processing null code here.
+         */
+        if (exitCode === null) {
+            return;
```
Comment (alexander-smolyakov) [inline Tasks/BashV3/bash.ts:165]: Same point, please add some log-message.

File: Tasks/CmdLineV2/cmdline.ts [hunk 4]
```diff
@@ -65,11 +69,23 @@ async function run() {
             });
         }
 
+        process.on("SIGINT", () => {
+            bash.killChildProcess();
```
Comment (alexander-smolyakov) [inline Tasks/CmdLineV2/cmdline.ts:73]: Same point, please add some log-message.

File: Tasks/CmdLineV2/cmdline.ts [hunk 5]
```diff
@@ -65,11 +69,23 @@ async function run() {
...
         let exitCode: number = await bash.exec(options);
 
         let result = tl.TaskResult.Succeeded;
 
+        /**
+         * Exit code null could appeared in situations if executed script don't process cancellation signal,
+         * as we already have message after operation cancellation, we can avoid processing null code here.
+         */
+        if (exitCode === null) {
+            return;
```
Comment (alexander-smolyakov) [inline Tasks/CmdLineV2/cmdline.ts:86]: Same point, please add some log-message.

**Review by alexander-smolyakov — COMMENTED** [review]
Could you please take a look at the comments? Other changes LGTM.

File: Tasks/BashV3/bash.ts (hunk 2 above)
Comment (EzzhevNikita) [reply Tasks/BashV3/bash.ts:133]: Thanks, I've added

File: Tasks/BashV3/bash.ts [hunk 6]
```diff
@@ -128,6 +128,11 @@ async function run() {
             ignoreReturnCode: true
         };
 
+        process.on("SIGINT", () => {
+            console.log('Started cancellation of executing script');
```
Comment (alexander-smolyakov) [inline Tasks/BashV3/bash.ts:132]: Thanks for adding log messages, but could you please use ` tl.debug` for output instead of ` console.log`?

**Review by alexander-smolyakov — APPROVED** [review]

**Comment by EzzhevNikita** [comment]
@damccorm Could you please take a look this pr?

File: Tasks/BashV3/bash.ts [hunk 7]
```diff
@@ -79,7 +79,7 @@ async function run() {
             if (old_source_behavior) {
                 contents = `. '${targetFilePath.replace(/'/g, "'\\''")}' ${input_arguments}`.trim();
             } else {
-                contents = `bash '${targetFilePath.replace(/'/g, "'\\''")}' ${input_arguments}`.trim();
+                contents = `exec bash '${targetFilePath.replace(/'/g, "'\\''")}' ${input_arguments}`.trim();
```
Comment (damccorm) [inline Tasks/BashV3/bash.ts:82]: Why do we need to exec here? What's the purpose of this change?

File: Tasks/CmdLineV2/cmdline.ts [hunk 8]
```diff
@@ -13,6 +13,10 @@ async function run() {
         let failOnStderr = tl.getBoolInput('failOnStderr', false);
         let script: string = tl.getInput('script', false) || '';
         let workingDirectory = tl.getPathInput('workingDirectory', /*required*/ true, /*check*/ true);
+        
+        if (fs.existsSync(script)) {
+            script = `exec ${script}`;
+        }
```
Comment (damccorm) [inline Tasks/CmdLineV2/cmdline.ts:19]: Same question here - why do we need this?

File: Tasks/BashV3/bash.ts (hunk 7 above)
Comment (EzzhevNikita) [reply Tasks/BashV3/bash.ts:82]: This change allows to the script to catch signals that are passed from toolRunner. Exec allows to user script file to be executed as subprocess of our Bash and CmdLine tasks. As we are wraping input scripts into temp script file, execution of user's scripts is starting as subprocess of the process where this temp script is executing and as a result user's script couldn't catch cancellation signal from task.

**Review by damccorm — APPROVED** [review]
